miner/Miner.hs

```diff
@@ -208,107 +190,114 @@ main = do
+    updateSignal :: RIO Env ()
+    updateSignal = catchAny f $ \_ -> do
+        logWarn "Couldn't connect to update stream. Trying again..."
+        updateSignal
       where
+        f :: RIO Env ()
+        f = do
+            e <- ask
+            liftIO $ withEvents (req $ args e) (mgr e) (void . SP.head_ . SP.filter realEvent)
+            cid <- liftIO chain
+            logInfo $ cid <> ": Current work was preempted."
+
         -- TODO Formalize the signal content a bit more?
         realEvent :: ServerEvent -> Bool
-        realEvent (ServerEvent _ _ _) = True
+        realEvent ServerEvent{} = True
         realEvent _ = False
 
         -- TODO This is an uncomfortable URL hardcoding.
-        req :: Request
-        req = defaultRequest
-            { host = B.pack . baseUrlHost . _url $ coordinator a
+        req :: ClientArgs -> Request
+        req a = defaultRequest
+            { host = encodeUtf8 . T.pack . baseUrlHost $ coordinator a
             , path = "chainweb/0.0/" <> encodeUtf8 (toText $ version a) <> "/mining/updates"
-            , port = baseUrlPort . _url $ coordinator a
+            , port = baseUrlPort $ coordinator a
             , secure = True
             , method = "GET"
-            , requestBody = RequestBodyBS $ _chainBytes cbytes }
+            , requestBody = RequestBodyBS cbs }
 
     -- | If the `go` call won the `race`, this function yields the result back
     -- to some "mining coordinator" (likely a chainweb-node). If `updateSignal`
     -- won the race instead, then the `go` call is automatically cancelled.
     --
-    miningSuccess :: HeaderBytes -> IO ()
+    miningSuccess :: HeaderBytes -> RIO Env ()
     miningSuccess h = do
-      bh <- bytesToBlockHeader h
-      putStrLn $ "Success!!!  " <> (show $ _blockNonce bh)
-      putStrLn $ "difficulty = " <> (T.unpack $ showTargetHex $ _blockTarget bh)
-      putStrLn $ "blockHash  = " <> (T.unpack $ hashToHex $ _blockHash bh)
-      putStrLn $ "powHash    = " <> (T.unpack $ powHashToHex $ _blockPow bh)
-      void . runClientM (solvedClient v h) $ ClientEnv m u Nothing
-      where
-        v = version $ args e
-        m = mgr e
-        u = _url . coordinator $ args e
-
-hashToHex :: BlockHash -> T.Text
-hashToHex = decodeUtf8 . B16.encode . runPutS . encodeBlockHash
-
-powHashToHex :: PowHash -> T.Text
-powHashToHex = decodeUtf8 . B16.encode . runPutS . encodePowHash
-
-bytesToBlockHeader :: HeaderBytes -> IO BlockHeader
-bytesToBlockHeader (HeaderBytes hbytes) = runGet decodeBlockHeaderWithoutHash hbytes
-
-cpu :: ChainwebVersion -> CPUEnv -> MWC.GenIO -> TargetBytes -> HeaderBytes -> IO HeaderBytes
-cpu v e g tbytes hbytes = fmap head . withScheduler comp $ \sch -> do
-    bh <- bytesToBlockHeader hbytes
-    let BlockHeight height = _blockHeight bh
-    printf "Mining block %d on chain %s with difficulty %s\n"
-      height
-      (T.unpack $ chainIdToText $ _blockChainId bh)
-      (T.unpack $ showTargetHex (_blockTarget bh))
```

Comment:
Doing all these here would slow down mining.

Reply:
This is a few instructions once every time a new block gets mined.  This is so far down in the weeds it's not worth worrying about.  Also, the instructions for doing these conversions will be dwarfed by the IO itself.